### PR TITLE
Support `$redirect` in DNR

### PR DESCRIPTION
**Pull request by seia-soto**
fixes https://github.com/ghostery/ghostery-extension/issues/1714

This PR extracts resources from the `ads` engine binary on build time and registers into manifest.

TODO

- [x] Complete base64 encoded binary decoding
- [x] Convert uBO resource names to adguard format (we use tsurlfilter to convert networkfilters to dnr rules) // done on backend (not sure about the staging)
- [x] Custom filters remapping

### CODE REVIEW

File: extension-manifest-v3/scripts/build.js [hunk 1]
```diff
@@ -123,6 +124,43 @@ engines.forEach((engine) => {
   if (result.stderr) process.exit(1);
 });
 
+// extract resources
+shelljs.mkdir('-p', resolve(options.outDir, 'web_accessible_resources'));
+
+const seenWebAccessibleResources = new Set();
+
+FiltersEngine.deserialize(
```
Comment (chrmod) [inline extension-manifest-v3/scripts/build.js:132]: this procedure could be done in the engine download script. That way it wont have be re-created on every build, but only when the engine is changed

File: extension-manifest-v3/scripts/build.js [hunk 2]
```diff
@@ -174,6 +212,10 @@ if (manifest.permissions.includes('declarativeNetRequest') && argv.watch) {
   manifest.permissions.push('declarativeNetRequestFeedback');
 }
 
+manifest.web_accessible_resources.push({
+  resources: Array.from(seenWebAccessibleResources),
```
Comment (chrmod) [inline extension-manifest-v3/scripts/build.js:216]: how about adding a whole `web_accessible_resources` folder to all manifests? that way we wont have to change the manifest in the build time

File: extension-manifest-v3/scripts/build.js [hunk 3]
```diff
@@ -123,6 +124,43 @@ engines.forEach((engine) => {
...
+  }
+
+  const path = resolve(options.outDir, 'web_accessible_resources', key);
+
+  shelljs.mkdir(
+    '-p',
+    dirname(resolve(options.outDir, 'web_accessible_resources', path)),
+  );
+  writeFileSync(
+    resolve(options.outDir, 'web_accessible_resources', key),
```
Comment (chrmod) [inline extension-manifest-v3/scripts/build.js:159]: lets put those is some folder, eg. `web_accessible_resource/adblocker`

**Review by chrmod — CHANGES_REQUESTED** [review]

File: extension-manifest-v3/scripts/build.js (hunk 2 above)
Comment (seia-soto) [reply extension-manifest-v3/scripts/build.js]: Moved to `web_accessible_resources/statics`

File: extension-manifest-v3/scripts/build.js (hunk 1 above)
Comment (seia-soto) [reply extension-manifest-v3/scripts/build.js]: Moved.

File: extension-manifest-v3/scripts/build.js (hunk 3 above)
Comment (seia-soto) [reply extension-manifest-v3/scripts/build.js]: Moved to `web_accessible_resources/statics`

File: extension-manifest-v3/src/manifest.chromium.json [hunk 4]
```diff
@@ -125,6 +125,11 @@
       "matches": [
         "<all_urls>"
       ]
+    },
+    {
+      "resources": [
+        "rule_resources/statics/*"
```
Comment (chrmod) [inline extension-manifest-v3/src/manifest.chromium.json:131]: should be merged with the list above as it is meant to be used on all_frames and with <all_urls>

Comment (seia-soto) [reply extension-manifest-v3/src/manifest.chromium.json:131]: We have a conflict with https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/scripts/build.js#L219-L242

File: extension-manifest-v3/scripts/download-engines/index.js [hunk 5]
```diff
@@ -140,3 +140,53 @@ for (const [name, target] of Object.entries(DNR)) {
...
+];
+
+FiltersEngine.deserialize(
+  readFileSync(`${TARGET_PATH}/engine-ads.dat`),
+).resources.resources.forEach((value, key) => {
+  // refs https://github.com/gorhill/uBlock/tree/master/src/web_accessible_resources
+  if (
+    value.contentType === 'application/javascript' &&
+    (value.body.includes('scriptletGlobals') || // Drop scriptlets
+      key.indexOf('/') !== -1) // Drop resources within a directory
```
Comment (philipp-classen) [inline extension-manifest-v3/scripts/download-engines/index.js:169]:
```suggestion
      key.includes('/')) // Drop resources within a directory
```

**Comment by chrmod** [comment]
@seia-soto please provide test cases for @GRadziejewski

File: extension-manifest-v3/src/manifest.chromium.json [hunk 6]
```diff
@@ -125,6 +125,11 @@
       "matches": [
         "<all_urls>"
       ]
+    },
```
Comment (smalluban) [inline extension-manifest-v3/src/manifest.chromium.json:128]: Is this only chromium feature? If not, all of the manifests should be updated.

**Review by smalluban — CHANGES_REQUESTED** [review]

File: extension-manifest-v3/src/manifest.chromium.json (hunk 6 above)
Comment (seia-soto) [reply extension-manifest-v3/src/manifest.chromium.json]: Fixed in [`0097ddf` (#1750)](https://github.com/ghostery/ghostery-extension/pull/1750/commits/0097ddf6d60ee01382125300b57ea596b7203bbf). Also, manifest changes were reverted to keep the build process simple.

**Review by smalluban — APPROVED** [review]

**Comment by seia-soto** [comment]
Hi @GRadziejewski ,

Here's a testing instruction for Ghostery Browser Extension.

Before we start, the purpose of `$redirect` option in filters is to avoid detection of ad-blocking and website breakage from the request failures. Also, we ship some mocked javascript files not to be detected. For example, we can inject Google Tag Manager shim instead of just redirecting to empty javascript file if the website is coded to require GTM API.

refs https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#redirect

I'm going to use the following website to test if we can make sure if `$redirect` works:

```
https://www.iana.org/help/example-domains
```

using the following filter:

```adb
||iana.org/_img/2022/iana-logo-header.svg$redirect=32x32.png
```

1. Open the website: https://www.iana.org/help/example-domains
2. Apply the filter: `||iana.org/_img/2022/iana-logo-header.svg$redirect=32x32.png`
3. Open dev tools then look for `iana-logo-header.svg` and check if its type is set to `/ Redirect` (depending on the browser, the message can be different)

Unlike the general request blocking, you always should get the response code of `30x redirect` then `200 ok`. If you see the request is blocked by `(blocked:others)` or `blocked by [extension name]` (in Firefox), we regard it as "redirect modifier failure".

**Comment by seia-soto** [comment]
Converting to draft: We have add a support for custom filters.

**Comment by seia-soto** [comment]
Pending update from https://github.com/ghostery/urlfilter2dnr/pull/16

**Review by smalluban — CHANGES_REQUESTED** [review]
The redirect locations don't match DNR rules...

For example, the Google Tag redirect shows me in the console:
`
chrome-extension://jccobfekniccmgppgbkonmdmdldhhgjc/rule_resources/redirects/googletagservices-gpt.js`

But the real location is `extension-manifest-v3/dist/rule_resources/redirects/googletagservices_gpt.js` (by the transfer in the build step). 

The `-` from DNR rule is `_` in the file system.

**Review by chrmod — APPROVED** [review]

**Comment by chrmod** [comment]
The redirect resource urls were fixed on the ruleset generation side.

**Review by smalluban — APPROVED** [review]

**Pull request merged**